### PR TITLE
JARVIS-968: validate skill tool inputs and harden document-editor flow

### PR DESCRIPTION
**Pull request by alex-nork**
## Summary

Closes JARVIS-968 (`document_editor` flow created duplicate docs after a malformed `document_update({})` call).

The feature branch implements end-to-end hardening of the document-editor flow:

- **Skill-tool factory layer:** new `validateInputAgainstSchema` utility rejects missing required fields, wrong types, and bad enum values for every skill tool's input BEFORE the executor runs, with self-correcting agent-readable error messages (`Invalid input for tool "..."`).
- **Executor layer:** `executeDocumentUpdate` (and sibling `executeDocument*` fns) defensively validate `surface_id`, `content`, and `mode` before any DB lookup, so a `document_update({})` call now surfaces as `Invalid input: surface_id is required …` instead of a misleading `Document not found`.
- **SKILL.md guidance:** new `Recovering from a failed update` subsection redirects the model from "create another doc" to "retry with the surface_id from the prior create call".
- **Server-side dedupe:** `executeDocumentCreate` now reuses an empty same-title document in the same conversation (within a 5-minute window) instead of creating a duplicate row.
- **macOS client safety:** `handleDocumentEditorUpdate` now gates on `documentManager.surfaceId == msg.surfaceId`, bringing macOS to parity with the web client and preventing the dedupe path from corrupting an unrelated active document.

## Self-review result

**PASS** after 3 rounds. Two remediation cycles addressed:
- **R1:** Codex P1 client/server race (closeDocument→save clobbering server content), document_conversations junction-row parity, mode=null inconsistency between factory validator and executor.
- **R2:** Wrong-doc data corruption regression introduced by R1's `_show → _update` swap — the macOS `_update` handler didn't gate on `surfaceId`, so the dedupe path could overwrite an unrelated active document and persist it via autoSave.

## PRs merged into feature branch
- #32375: docs(document-editor): clarify update-failure recovery in SKILL.md
- #32376: feat(assistant): add JSON schema input validator for skill tools
- #32377: fix(document-editor): validate update input before lookup
- #32382: fix(document-editor): reuse empty same-title doc in same conversation
- #32385: feat(assistant): enforce required/type/enum on skill tool inputs

### Fix PRs
- #32390: fix(document-editor): harden dedupe path and align mode null handling
- #32394: fix(macos): gate document_editor_update on surface_id match

## Deferred follow-ups (not blocking)

- **Recurse `validateInputAgainstSchema` into array `items`** for `enum`/`required` checks. Codex P2 on PR #32376 flagged that bundled skills like `messaging_list_conversations.types.items.enum` and `sequence_create.steps.items.required` aren't enforced by the new validator. Plan explicitly scoped this out ("strict subset of JSON Schema so we never reject a legitimate call"); existing executor-side checks still catch the bad data. Worth a follow-up enhancement if we want defense-in-depth at the validator layer.

Part of plan: doc-editor-validate.md

### CODE REVIEW

**Comment by alex-nork** [comment]
<details>
<summary>Plan papertrail: JARVIS-968: Validate skill tool inputs and harden document-editor flow</summary>

# JARVIS-968: Validate skill tool inputs and harden document-editor flow

## Overview
JARVIS-968 captures a user feedback bundle where the document-editor skill
created a duplicate document after the model issued `document_update({})`. The
runtime only validates unknown parameters and lets a malformed call reach the
executor, which then returns the misleading "Document not found". This plan
adds full JSON-schema validation for skill tool inputs, hardens
`executeDocumentUpdate` with defensive checks and clearer errors, tightens the
SKILL.md recovery guidance, and adds a dedupe guard so a same-title empty
document in the current conversation is reused instead of duplicated.

## PR 1: Add validateInputAgainstSchema utility
### Depends on
None

### Branch
doc-editor-validate/pr-1-validate-input-against-schema

### Title
feat(assistant): add JSON schema input validator for skill tools

### Files
- assistant/src/skills/validate-input.ts
- assistant/src/__tests__/validate-input.test.ts

### Implementation steps
1. Create `assistant/src/skills/validate-input.ts` exporting one function:
   ```ts
   export interface InputValidationFailure {
     ok: false;
     errors: string[]; // human-readable messages, e.g. 'surface_id is required'
   }
   export interface InputValidationSuccess { ok: true }
   export type InputValidationResult = InputValidationSuccess | InputValidationFailure;
   export function validateInputAgainstSchema(
     toolName: string,
     input: Record<string, unknown>,
     schema: Record<string, unknown> | undefined,
   ): InputValidationResult
   ```
2. Cover the JSON-schema features actually used across
   `assistant/src/config/bundled-skills/**/TOOLS.json`:
   - `required: string[]` — each listed key must exist in `input` (use `in`
     check; do not treat `undefined`/`null` as present).
   - `properties[k].type` for `string`, `number`, `integer`, `boolean`,
     `array`, `object`. For `integer`, additionally require
     `Number.isInteger(value)`.
   - `properties[k].enum` for string enums (e.g. `mode: "replace" | "append"`).
   - `properties[k].items.type` for arrays — verify each element matches the
     declared item type.
   - Unknown keys (parity with the existing `validateNoUnknownParams`):
     anything in `input` not present in `properties`.
3. Skip validation entirely when `schema` is null/undefined or has no
   `properties` block, matching today's lenient behavior (covered by existing
   skill-tool-factory tests).
4. Do **not** attempt to support `$ref`, `oneOf`, `anyOf`, `allOf`, `format`,
   `pattern`, `minimum`, `maximum`, `additionalProperties: object-shape`. If
   the schema contains these keywords, simply skip validation of that
   sub-schema (do not throw). Keep the validator a strict subset of JSON
   Schema so we never reject a legitimate call.
5. Each error message must be agent-readable and self-correcting. Examples:
   - `surface_id is required`
   - `content must be a string`
   - `mode must be one of "replace", "append"`
   - `Unknown parameter "foo". Supported: "surface_id", "content", "mode"`
6. Write `assistant/src/__tests__/validate-input.test.ts` covering:
   - Required-field success and per-missing-field failure (lists each missing
     field individually).
   - Type checks for `string`, `number`, `integer`, `boolean`, `array`,
     `object` (success + failure paths).
   - Enum success and enum failure.
   - Array `items.type` success and per-element failure.
   - Unknown-keys detection (single and multiple).
   - Empty `input` is allowed when nothing is `required`.
   - Schema with no `properties` / null schema is permissive.
   - Schemas containing `oneOf` / `$ref` do not throw (treated as
     "skip sub-schema").

### Acceptance criteria
- `bun test assistant/src/__tests__/validate-input.test.ts` passes.
- No existing tests are affected (this PR adds a standalone module).
- The validator function is pure (no I/O, no mutation of inputs).

## PR 2: Defensive validation in executeDocumentUpdate
### Depends on
None

### Branch
doc-editor-validate/pr-2-document-update-defensive

### Title
fix(document-editor): validate update input before lookup

### Files
- assistant/src/tools/document/document-tool.ts
- assistant/src/__tests__/document-tool-security.test.ts (extend) OR new file `assistant/src/__tests__/document-update-validation.test.ts`

### Implementation steps
1. In `assistant/src/tools/document/document-tool.ts`, add a small helper
   above `executeDocumentUpdate`:
   ```ts
   function invalidInput(message: string): ToolExecutionResult {
     return {
       content: JSON.stringify({ success: false, error: `Invalid input: ${message}` }),
       isError: true,
     };
   }
   ```
2. Replace the first three lines of `executeDocumentUpdate` (currently
   `const surfaceId = input.surface_id as string;` etc.) with explicit
   validation:
   - If `typeof input.surface_id !== "string"` or empty after trim →
     `return invalidInput("surface_id is required and must be a non-empty string");`
   - If `typeof input.content !== "string"` →
     `return invalidInput("content is required and must be a string");`
     (Allow empty string — appending nothing is a valid no-op.)
   - If `input.mode !== undefined && input.mode !== "replace" && input.mode !== "append"` →
     `return invalidInput('mode must be "replace" or "append"');`
   Only after these checks, proceed to `canAccessDocument` / `getDocumentById`.
3. Apply the same defensive `surface_id` check to the other
   `executeDocument*` functions that cast `input.surface_id as string`
   without validation: `executeDocumentOpen`, `executeDocumentRead`,
   `executeDocumentDelete`, `executeDocumentFind`,
   `executeDocumentReplaceText`. Each should bail out with
   `invalidInput("surface_id is required and must be a non-empty string")`
   before any DB / access work. Do not change behaviour for `executeDocumentCreate`
   or `executeDocumentList` (their schemas have no required fields).
4. Extend `assistant/src/__tests__/document-tool-security.test.ts` with a
   new `describe("executeDocumentUpdate — input validation", ...)` block
   covering:
   - `executeDocumentUpdate({}, ctx)` returns
     `{ isError: true, content: includes "Invalid input: surface_id is required" }`
     and never reads `Document not found`.
   - `executeDocumentUpdate({ surface_id: "doc-x" }, ctx)` returns
     `Invalid input: content is required`.
   - `executeDocumentUpdate({ surface_id: "doc-x", content: "" }, ctx)`
     does NOT fail validation (empty string is allowed) — it falls through
     to the access check.
   - `executeDocumentUpdate({ surface_id: "doc-x", content: "hi", mode: "bogus" }, ctx)`
     returns `Invalid input: mode must be "replace" or "append"`.
   - Add one parallel test for `executeDocumentRead({}, ctx)` and one for
     `executeDocumentDelete({}, ctx)` to lock in the cross-tool change.

### Acceptance criteria
- `bun test assistant/src/__tests__/document-tool-security.test.ts` passes.
- A malformed `document_update` call surfaces as `Invalid input: ...`
  instead of `Document not found`.
- No regression: existing tests that pass a real `surface_id` still
  succeed.

## PR 3: Wire validateInputAgainstSchema into skill-tool-factory
### Depends on
PR 1

### Branch
doc-editor-validate/pr-3-skill-tool-schema-validate

### Title
feat(assistant): enforce required/type/enum on skill tool inputs

### Files
- assistant/src/tools/skills/skill-tool-factory.ts
- assistant/src/__tests__/skill-tool-factory.test.ts

### Implementation steps
1. In `assistant/src/tools/skills/skill-tool-factory.ts`, delete the local
   `validateNoUnknownParams` helper. Import the new utility:
   ```ts
   import { validateInputAgainstSchema } from "../../skills/validate-input.js";
   ```
2. Replace the validation block inside the inline `execute` function
   (currently lines 69-74) with:
   ```ts
   const validation = validateInputAgainstSchema(
     entry.name,
     input,
     entry.input_schema as Record<string, unknown> | undefined,
   );
   if (!validation.ok) {
     return {
       content: `Invalid input for tool "${entry.name}": ${validation.errors.join("; ")}. Fix the arguments and retry.`,
       isError: true,
     };
   }
   ```
3. Update `assistant/src/__tests__/skill-tool-factory.test.ts`:
   - Keep the existing "unknown parameter validation" describe block,
     but adjust the expected error string format if necessary (search for
     `Unknown parameter` and update). The new message uses
     `Invalid input for tool "test_tool": Unknown parameter "unsubscribe". ...`.
   - Add a new `describe("createSkillTool — required/type/enum validation")`
     block covering:
     - Missing required field returns `isError: true` with message
       containing `query is required` and `Invalid input for tool`.
     - Wrong type (`query: 123` when schema declares string) returns
       `isError: true` with `query must be a string`.
     - Enum violation (schema with `mode: { enum: ["a","b"] }`,
       input `{ mode: "c" }`) returns
       `isError: true` with `mode must be one of "a", "b"`.
     - Valid input passes through to the executor unchanged (already
       covered, but assert the value reaches `echo.ts`).
4. Manually verify in a smoke test that the bundled
   `document_update` tool now rejects `{}` at the factory layer (no need to
   add a new test — the document-tool defensive checks in PR 2 already
   cover this end-to-end at the executor layer, and PR 3 ensures the
   schema check fires first for any skill tool).

### Acceptance criteria
- `bun test assistant/src/__tests__/skill-tool-factory.test.ts` passes.
- `bun test assistant/src/__tests__/` (full suite) passes — no regressions
  in other skill-tool tests.
- `document_update({})` now fails with an `Invalid input for tool
  "document_update": surface_id is required; content is required` style
  message at the factory layer, before reaching `executeDocumentUpdate`.

## PR 4: Document-editor recovery guidance in SKILL.md
### Depends on
None

### Branch
doc-editor-validate/pr-4-skill-md-recovery

### Title
docs(document-editor): clarify update-failure recovery in SKILL.md

### Files
- assistant/src/config/bundled-skills/document-editor/SKILL.md

### Implementation steps
1. In `assistant/src/config/bundled-skills/document-editor/SKILL.md`, add
   a new subsection under the existing `## Creating a new document` section,
   right after step 3 (the "Stream content in chunks" step), titled
   `### Recovering from a failed update`:
   ```
   ### Recovering from a failed update

   If a `document_update` call fails with an `Invalid input` error
   (for example because `surface_id` was missing), do NOT call
   `document_create` again. The `surface_id` you need is in the
   tool result of the most recent `document_create` call in this turn.
   Retry `document_update` with that `surface_id` and the same content.
   Creating a second document with the same title produces a duplicate
   for the user.
   ```
2. Tighten step 1 of `## Creating a new document` to reinforce the surface
   id contract: append the sentence `Capture the `surface_id` from the
   response — every subsequent `document_update` call must reference it.`
3. Do not modify TOOLS.json or any executor in this PR.

### Acceptance criteria
- The SKILL.md change is text-only; `bun test` still passes.
- The new subsection is reachable when the skill is loaded
  (bundled-skills are loaded directly from disk — no separate build step).

## PR 5: Dedupe empty same-title documents in same conversation
### Depends on
PR 2

### Branch
doc-editor-validate/pr-5-document-create-dedupe

### Title
fix(document-editor): reuse empty same-title doc in same conversation

### Files
- assistant/src/documents/document-store.ts
- assistant/src/tools/document/document-tool.ts
- assistant/src/__tests__/document-tool-security.test.ts

### Implementation steps
1. In `assistant/src/documents/document-store.ts`, add a new exported
   helper near `searchDocumentsByTitle`:
   ```ts
   /** Return the most recent empty document in the given conversation with
    *  the supplied title, created within the last `withinMs` milliseconds.
    *  Used to dedupe a duplicate create-then-create flow after a failed update. */
   export function findRecentEmptyDocumentByTitle(
     conversationId: string,
     title: string,
     withinMs: number,
   ): { surfaceId: string } | null
   ```
   Implement it as a single SELECT on the existing `documents` table:
   ```
   SELECT surface_id FROM documents
   WHERE conversation_id = ?
     AND title = ?
     AND content = ''
     AND created_at >= ?
   ORDER BY created_at DESC
   LIMIT 1
   ```
   Compute `created_at` threshold as `Date.now() - withinMs`.
2. In `assistant/src/tools/document/document-tool.ts`, modify
   `executeDocumentCreate`:
   - After resolving `title` and `initialContent`, but BEFORE generating
     a new `surface_id`, call `findRecentEmptyDocumentByTitle(
     context.conversationId, title, 5 * 60 * 1000)` (5-minute window).
   - If a match is found AND the incoming `initialContent` is non-empty,
     reuse the existing `surfaceId` instead of creating a new row. Call
     `updateDocumentContent(existingSurfaceId, initialContent, "replace")`
     and send the same `document_editor_show` + `ui_surface_show` +
     return payload (with `reused: true` added to the JSON content and
     `message: "Document editor reopened (deduped empty draft)"`).
   - If `initialContent` is empty, fall through to the existing
     create-new-row path (no dedupe — the model likely intends a fresh
     empty doc).
   - Place the dedupe logic behind an internal helper
     `function maybeReuseEmptyDocument(title, initialContent, context)`
     returning `ToolExecutionResult | null`.
3. Import `findRecentEmptyDocumentByTitle` in `document-tool.ts`.
4. Extend `assistant/src/__tests__/document-tool-security.test.ts` (or a
   new file `document-create-dedupe.test.ts`) with:
   - Seed an empty document `{ title: "My Post", content: "", created_at: now }`
     in `conv-current`.
   - Call `executeDocumentCreate({ title: "My Post", initial_content: "Hello" }, ctx)`.
   - Assert: result is `success: true`, returned `surface_id` matches the
     seeded one, `reused: true` is present in the JSON.
   - Assert: `getDocumentById(seededId).content === "Hello"`.
   - Assert: `getDocumentsForConversation("conv-current").length === 1`
     (no duplicate row).
   - Add a negative test: empty seeded doc but `initial_content === ""`
     in the incoming call → a new doc is created (no dedupe).
   - Add a negative test: seeded doc with non-empty content → no dedupe,
     a new doc is created.
   - Add a negative test: seeded empty doc older than the 5-minute window
     → no dedupe.

### Acceptance criteria
- `bun test assistant/src/__tests__/` passes.
- After the malformed-update scenario from JARVIS-968, the recovery path
  reuses the empty draft rather than producing a second same-title
  document.
- The dedupe only triggers for empty documents in the same conversation
  within the 5-minute window; all other create paths behave exactly as
  before.

</details>

File: assistant/src/skills/validate-input.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+  // 2. Per-property checks: type, enum, items.type.
+  for (const [key, rawSubSchema] of Object.entries(properties)) {
+    if (!(key in input)) continue;
+    const value = input[key];
+    // `undefined` / `null` are treated as absent for type-checking purposes
+    // so we never reject a schema that legitimately permits null (e.g. a
+    // plugin schema with `type: ["string","null"]`). Combined with the
+    // presence-only `required` check above, this preserves nullable inputs.
+    if (value === undefined || value === null) continue;
```
[the remark this answers is not in the document]
Comment (alex-nork) [reply assistant/src/skills/validate-input.ts]: Addressed: single-type schemas now reject null at the factory layer; union types still bypass.

**Comment by alex-nork** [comment]
@codex review this PR again — the previous P2 finding (null acceptance for non-nullable schema types) has been fixed in commit 0e2f8cdca0

**Comment by alex-nork** [comment]
@devin review this PR again — the previous P2 finding (null acceptance for non-nullable schema types) has been fixed in commit 0e2f8cdca0

**Comment by alex-nork** [comment]
@codex review this PR again — fix in 0e2f8cdca0 addressed the P2 finding. Please confirm.
